### ISSUE TITLE
Plans for future of "spec-like" API

### ISSUE THREAD

**Issue by danielballan** [NSLS-II/bluesky#677]
This issue documents our intentions for unifying some features of the "spec-like" API with the "standard" built-in plans and any user-defined plans. The motivation of this change is to avoid two disjoint forks of plans development.

For background, see [differences from non-spec-like plans](http://nsls-ii.github.io/bluesky/plans.html#differences-from-non-spec-like-plans).

## 1. Smarter Dispatcher

Motivation: Make it easier to apply nice subscriptions (e.g. plots) to all plans where they apply

Enhancement: Refactor the dependency-injection scheme in `bluesky.plans.spec_api` into a  "smart" Dispatcher-like object that can apply to *all* plans (including built-in standard plans, user-defined plans, and spec-like plans) auto-subscribing relevant subscriptions based on the contents of the RunStart document and/or EventDescriptors.

Planned Release: v0.9.0 (next release)

```python
# Currrent
RE(bluesky.spec_api.ct(...))  # provides table
RE(bluesky.plans.count(...))  # provides nothing

# Planned
RE(bluesky.spec_api.ct(...))  # provides table
RE(bluesky.plans.count(...))  # provides nothing
D = bluesky.default_dispatcher()
RE.subscribe('all', D)
RE(bluesky.plans.count(...))  # now provides table
```

`D` will contain logic for mapping metadata to appropriate subscriptions, such as 1D plots for 1D scans and various 2D plots for various types of 2D scans. It will provide methods to configure its behavior interactively, including an 'off' switch. It will absorb many of the attributes that we currently hang off of `gs`, such as `PLOT_Y` and `TABLE_COLS` (maybe). It does _not_ include `gs.DETS`. See below for the fate of `gs.DETS`.

## 2. Consistent Signature and Deprecating Global State Detectors

Motivation: In an early version, we introduced the global detectors list `gs.DETS` as a way to synchronize plans and automatic subscriptions. Now that plans report richer metadata, it's not necessary, and the role of `gs.DETS` can be replaced by a normal user-defined variable. This simplifies the bluesky codebase and the usage, and it allows users to switch between multiple different lists easily.

API break: Require the detector argument for spec-like plans.

Planned Release: v0.10.0

```python
# Current
gs.DETS = [detector1, detector2]
ct()  # implicitly uses gs.DETS for detectors

# Planned
d = [detector1, detector2]
ct(d)  # where `d` is a list of detectors, akin to how some users use `gs.DETS` now
# The old usage, `ct()` with no arguments, is an error!
```

**Comment by danielballan** [comment]
Update/elaboration: In the next release of bluesky...

* Beamlines will be encouraged to subscribe a `BestEffortCallback` to the RunEngine, providing a simple table and multi-axis plot by default for _all_ plans.
    ```python
    from bluesky.callbacks import BestEffortCallback
    B = BestEffortCallback()
    RE.subscribe('all', B)
    # supports B.disable() and B.enable() for easy on/off
    ```
* Beamlines will be encouraged to add a `DiagnosticPreprocessor` to the RunEngine's list of plan preprocessors. Here, users can set up lists of devices to monitor or fly in the background during a run, as well set the list of devices to take 'baseline' readings of before and after a scan.
    ```python
    from bluesky import DiagnosticPreprocessor
    D = DiagnosticPreprocessor(monitors=[beam_current],
                               flyers=[diag_cam],
                               baseline=[motor1, det3])
    # supports updating/inspection via D.monitors, D.flyers, D.baseline
    RE.preprocessors.append(D)
    ```
* No longer having a purpose, `bluesky.spec_api` will be deleted entirely. Beamlines may alias the standard plans with SPEC-like names _if they wish_ (`from bluesky.plans import count as ct`) once they take note of the different API. This ends the bad situation of having two incompatible branches of plan development.
* Likewise, `bluesky.global_state` and its `gs` object will be deleted entirely. Its purpose is better served by the DiagnosticPreprocessor and the BestEffortCallback.

(To be clear, at the time of this post, neither `BestEffortCallback` or `DiagnosticPreprocessor` have been submitted into bluesky as a PR just yet, but they have been prototyped at SIX and seem nice to use.)

**Comment by stuartcampbell** [comment]
I've been thinking about the future of the `spec_api` plans, and I know @danielballan is in favour of removing them, but I'm a little worried that if we don't define them then we might end up with many many non-standard names for the plans.  So I would be in favour of defining the `spec_api` plans names, but make them just point to (or call) the standard builtin plans.  Thoughts ?

**Comment by danielballan** [comment]
This was actually my original plan, as in:

```python
# bluesky/spec_api.py
from .plans import (
    count as ct,
    scan as ascan
    # etc.
)
```

If, instead, we removed `spec_api.py` entirely and prompted BL scientists to add that statement to their IPython profile, maybe that would help highlight for them that (1) these are just aliases now and (2) critically, the API has changed. Bluesky will have removed the functions entirely; it will be up to users if they want to reuse the old names.

In practice I don't think it will make much difference to the BL staff. I don't have strong feelings about which approach is best.

**Comment by prjemian** [comment]
This is a reasonable approach.  We spend additional maintenance time supporting code that tries to make one application the same as another.  Best to make this SPEC-like interface have some cost to the user.

**Comment by stuwilkins** [comment]
@danielballan are you assuming your code example above would be standard for NSLS-II?

**Comment by stuartcampbell** [comment]
I would be in favour of just maintaining a single set of aliases centrally in blue sky rather than trying to maintain and sync the separate configs on the beamlines

**Comment by stuartcampbell** [comment]
...for the spec commands that is (as they are common) and if a single beamline defines `ascan` as something different to spec it will only cause confusion.

**Comment by stuwilkins** [comment]
:+1: @stuartcampbell , we are starting to have beamline staff move between beamlines and a consistency is really important. As long as `ascan` does the same on all beamlines I think this is a good solution.

**Comment by stuwilkins** [comment]
This may be off topic, but I feel some of the issues surrounding the "spec_api" are that it is related to "spec". Some people love it ... some people hate it because of its association with a certain c-based acquisition software. 

I think if you boil this down the "requirement" is to have a base set of plans that do the same on all beamlines and are simple to use. I personally think these should be standard in `bluesky` not defined in a beamlines `ipython_profile`.

We have become used to a certain command set such as `ascan` and keeping that constant is important. Asking people to change from typing `ascan` to something else just to do an `ascan` is a bit like asking linux users to stop using `ls` ........ 🦆

**Comment by danielballan** [comment]
I'm hearing a convincing argument for shipping standard plans (or plan aliases) that take their names from SPEC.

Following up @stuwilkins's comment, "We have become used to a certain command set such as `ascan` and keeping that constant is important," I want to highlight that what we are proposing here will, effectively, be a major API change on `ct`, `ascan`, and friends. The current functions mimic SPEC's parameters as well as their names, indulging in non-idiomatic contortions to make that work. The new versions of `ct`, `ascan`, and friends would keep those names, but they would have different parameters.* Our motivation is to make bluesky overall simpler and self-consistent, to stop "trying to make one application the same as another," to borrow @prjemian's phrase.

We did plan for this possibility: we have intentionally never presented the  SPEC-like API in a tutorial, and most beamlines haven't taken it up. For those beamlines that have taken it up, we will send a very clear message about what has changed. It will be the most disruptive change to the library in several releases and the biggest API breakage we impose before declaring "1.0" and long-term stability.

*Examples:

```
# current
gs.DETS = [detector1, detector2]
ct()  # current

# planned
d = [detector1, detector2, ...]
ct(d)

# current  -- the only required argument is the *last* one which is unidiomatic
a2scan(motor1, start1, stop1, motor2, start2, stop2, num)

# planned  -- note changes order of 'num' parameter
a2scan(d, num, motor1, start1, stop1, motor2, start2, stop2)
```

**Comment by stuwilkins** [comment]
@danielballan I think that you perhaps misunderstood my comment, if so I am sorry I didn't explain it better. I am also not sure on your "most beamlines don't use it" I feel this is a scattering/spectroscopy .... potato/potato argument and I see that many of the similar beamlines I use doing the same......

What I said is:

- We should have a consistent set amongst all facility beamlines of a set of plans that do the same thing and are simple to use. What is to be avoided is that `ascan` has a different behaviour on 23-ID than on 2-ID (for example) 
- Wherever possible these standard plans should also have a base level of subscriptions such as `live_table` and `live_plot`, again so one person on 23-ID can use 2-ID (my current experiment!) and will see the same. 
- The API should be stable and not change unless there is some form of paradigm shift! 

Taking your example, I see only two changes:

- The removal of "global state" so you always have to add a list of detectors. Personally, I can see arguments for and against global state both as a programmer and a user. There is not enough space here, but I can see that removing global state to be good (for example to define multiple detector lists), so I come down in favor. 
- The moving of the num (is this number of points or number of intervals?) to the 2nd arg from the end however seems a weaker argument, but I suppose bundling all the changes together makes sense. I am not convinced on this one as num could always be arg[-1] to use the counter argument.

So, perhaps @danielballan we are in agreement! I only have a big concern about API churn this close to the user. Perhaps that is inevitable with this kind of development (which I might add is wonderful! Kudos to the bluesky team!)

Is this right?

**Comment by danielballan** [comment]
@stuwilkins We are on the same page. There are two kinds of "consistency" in this discussion -- consistency across beamlines and API consistency over time -- and I wanted to make sure we (or anyone reading along) weren't conflating them.

The overriding motivation for the API change is to unify the two disjoint forks of plans development, the one with global state and one without. We propose to do this by dropping global state because it is less flexible an arguably not any more convenient that using a simple global variable (in my example, `d`). Let's take it as a given that we are removing global state -- and uniformly adding a required `detectors` argument to all the SPEC-like plans to serve in its place -- and focus on a second aspect of this change.

I flagged `a2scan` because it's the most flagrant example of the mismatch between our SPEC-like function signatures and the signatures of "standard plans." Currently `a2scan` puts `num` _last_ (yes, using `args[-1]` as you describe) while the non-SPEC-like `inner_product_scan` puts `num` _first_. Two of our options going forward are:

1. Maintain `a2scan` as a thin wrapper around `inner_product_scan` that maintains the current API (modulo that new 'detectors' argument).

    ```python
    def a2scan(detectors, *args):
        num = args[-1]
        yield from inner_product_scan(detectors, num, args[:-1])
    ```

2. Make `a2scan` a simple alias for `inner_product_scan`. Suffer the `num`-related API change in exchange for greater self-consistency within bluesky -- but less "consistency" with SPEC.

I originally advocated for (2) but this discussion is pushing me toward (1). I feel comfortable asking our users to change to a no-global-state API. It requires one simple and consistent change of habits. But any additional changes, only in the name of internal self-consistency and simplifying our codebase, are probably not worth the hassle.

About my comment about "most beamlines haven't taken it up" -- of course nearly _everyone_ uses `count` and `scan` but only the beamlines that got bluesky very early on were taught about `ct` and `ascan`.

Thanks for engaging in this admittedly tedious but important discussion!

**Comment by stuwilkins** [comment]
@danielballan I am wondering it at this stage this reveals something that could do with a bit of thought. From what I see here there are two "forks" that happened during development. 

- Removal of "global state"
- Change of name of certain plans from things like `ct` to `count` and `mesh` to `outer_product_scan`

I am completely on board with dropping global state, global variables are just messy, make things difficult to debug. In fact, my current github punchcard is full of me doing this exact thing in `c` to a library I developed. I can even see many advantages, with careful setup of profiles one can make multiple detector lists etc. for different purposes instead of doing `gs.DETS.append` and `gs.DETS.remove` constantly. 

For the second, lets say for arguments sake that the `num` is the second argument. This means that `args[0]` and `args[1]` are detectors and num respectively.  Basically (as you suggested  above) then this becomes an alias list:

```
from .plans import (
    count as ct,
    scan as ascan,
    mesh as outer_product_scan,
    anscan as inner_product_scan, 
    ...
)
```

This gets me thinking, **_"what is the value in this"_**? 

Well, from what I see, the main problem (apart from _its not spec_) is that the names of the standard plans are perhaps not intuitive and too long. 

For example, there is a reason that`ls` in unix lists directory contents and not `list_directory_contents`. (And of course those of us _blessed_ to have used VMS remember `DIRECTORY` vs `ls`  and `DIRECTORY /FULL /TOTAL` vs `du` ) is because when you type it 200 times a day it is just a pain. 

Secondly, while I understand exactly why `outer_product_scan` is used in analogy with _linear algebra_ I do worry that while physicists understand linear algebra there are other disciplines at a facility that don't. 

So, I would suggest that if we agree on the first point (placement of args) that perhaps a small group of stakeholders could decide on definitive names for these standard plans and place some harmony on them. That way we get a consistent set, and we can harmonize across beamlines. I think this would address both the API and beamline consistency and would help to correctly name plans which in future make it into the standard set. In this point I worry a bit because if this whole thread has proved one thing, it is once a name gets out in the open, it is *very* *very* difficult to revert it........

**Comment by stuwilkins** [comment]
P.S. I wanted to add also, that I feel that the more non-intuitive we make standard plan names, the more likely it is to see aliases creep in in a non (facility/beamline) standard way.

**Comment by danielballan** [comment]
There a couple discussions we could spin off of this issue. Some might go more quickly in person, working with the "small group of stakeholders" you suggest -- and which @stuartcampbell has also raised.

I propose this way forward.

In the next release:

* **All plans get access to automatic plots and monitoring.** The addition of the DiagnosticPreprocessor and BestEffortCallback will make all the special aspects of `spec_api.py` plans unneeded. Users will be able to get automatic plotting, printing, monitoring, etc., for *all* plans without special-casing the SPEC-like ones.
* **The SPEC-like plans are maintained for now to minimal API disruption.** The plans in `spec_api.py` will become shims that merely alias the 'standard' plans or at most rearrange parameters. Thus, `a2scan` will still work (modulo the addition of the `detectors` arg).
* **Nudge users toward the standard plans.** The plans in `spec_api.py` will issue deprecation warnings, nudging users of to use the standard plans to reduce the number of names floating around.

In the next + 1 release:

* A group of stakeholders will revisit the names of the 'standard' plans. I don't think any API breakage will be needed here -- we can maintain old names forever to avoid breaking user code -- but we can change our minds about what names we choose to document and teach.

Keep in mind that the fundamental unit is the `Msg`, not the plan. Beamlines may want to build custom variants of things like `count` in order to cope with their specific needs. Of course, the more interoperable we can make these things, the better!

**Comment by stuwilkins** [comment]
So ... I think I have come full circle here. 

From where I sit, I would suggest removing the spec names on the next release % that the _automatic plots and monitoring_ works mostly like the current `spec_api.py`. I realise it won't be perfect, but my main worry from the beginning was that we were not just renaming and moving args but going much further up the stack and relying on more "at the beamline" customization. 

If the real changes are things like:

```
ct() --> count([fccd])
ascan(tth, start, stop, step) --> scan([fccd], tth, start, stop, steps)
dscan(tth, start, stop, step) --> relative_scan([fccd], tth, start, stop, step)
```
and the end user gets the same plots, table etc. **Then I think it is time to rip off the band-aid!** one more cycle of release does not buy anything. 

I do think at a later date a revisit the names is a good idea (and thanks @danielballan for considering it). Perhaps one way here is to have an alias list such as:

```
count -> ct
relative_scan --> rscan
outer_product_scan -> mesh
relative_outer_product_scan -> rmesh
```

Has some validity **if standardized** so all beamlines are the same and helps for the _command line warriors_ who use it a lot. Thoughts?

**Comment by danielballan** [comment]
I may have buried the lede: I _do_ intend to break the API in the next release by requiring the `detectors` arg for all plans that we ship. Maybe that was already clear -- if so, apologies for cudgeling with it. :- )

But, given that we currently have both `inner_product_scan` and `a2scan`, why confuse users by moving the `num` argument in `a2scan`? Why not continue to support the current argument order (modulo inserting `detectors`) and issue a warning?

**Comment by stuwilkins** [comment]
I think my point was that having an API break (by requiring the dets for the spec_api) then in the next cycle removing the spec_api altogether is just double the pain! People have to get used to a new order on a command and then once they are used to it we remove it! Like I said, if the real changes are just names and orders then I will just need to reprogram my muscle memory!

**Comment by danielballan** [comment]
Gotcha. Your point is, if users are going to have to fix their `a2scan` code anyway (because of `detectors`) we might as well force them to switch to `inner_product_scan` in one move -- rather than fix it, see the warning, fix it again. That's a reasonable position, and it's obviously easier for us! I'd be interested to know what @stuartcampbell and @tacaswell think.

**Comment by danielballan** [comment]
The specifications outlined in this issue were merged in #720, but there is some unresolved discussion. Continue in #740 and #741
